### ISSUE TITLE
Architecture question

### ISSUE THREAD

**Issue by nicolas-chaulet** [Jia-Research-Lab/PointGroup#7]
Hi, thanks a lot for the great work!
I am currently trying to implement that as part of [Torch Points3D](https://github.com/nicolas-chaulet/torch-points3d) and diving deeper into the architecture I was wondering why you use a UNet for the scoring module as opposed to a simple encoder. I tried with an encoder (simpler to implement and seems logical) but as soon as the score loss kicks in the model diverges and the score loss does not really decrease with time so I was wondering if it might be coming from that or something else.
Thanks!
Nicolas

**Comment by nicolas-chaulet** [comment]
Update: I have trained with a Unet architecture for the scorer and observe the same behaviour. If I cut the gradient before the scorer (the score loss does not impact the semantic segmentation network) then training is a lot more stable and seems to make sense. But obviously the features in the main unet cannot learn from the score loss.

**Comment by llijiang** [comment]
Hi Nicolas,

In my experience, the score loss usually decreases at first and then gradually increases, and finally tends to be stable. It is because there are many negative samples in proposals at first and the proposal quality is gradually getting better along with the training process. The scores are used to select better proposals and the performance could be much better with the scores in my experiment. 

And I think using a simple encoder as the score net should also make sense. I am not sure what causes your problem. A possible reason is that we train for several epochs before adding the score loss into the total loss. 

Hope it helps!

**Comment by nicolas-chaulet** [comment]
Thanks a lot for getting back to us! Yes, I followed your protocole as close as I could (we use Minkowski as our backbone). 
The grouping works well but the semantic branch goes crazy at the first epoch we activate our "scorer". It as if the gradients were so strong that it just destroys everything it has learned so far. Below are a few graphs that illustrate the phenomenum. On the blue and gray curve I propagate gradients all the way through the network whereas with the green curve, I cut the backpropagation of the scorer so that it does not affect the main backbone (full logs [here](https://app.wandb.ai/nicolas/panoptic?workspace=user-nicolas))
![image](https://user-images.githubusercontent.com/3178237/86597249-60010a00-bf93-11ea-8351-336dd30d74d8.png)

Having said that, could you please confirm that:

1. You don't have those crazy steps as soon as you activate the scorer
2. The gradients of the scrorer backpropagate all the way through the main backbone

really sorry to bother you with that but I am having trouble setting up spconv and wanted to double check some obvious things before I spend too much time trying to set that up..
Many thanks

**Comment by llijiang** [comment]
Hi, I checked my log again. For your questions,
1. Another losses are not obviously affected by the score loss in my experiments. The training curves are as follows. 
![image](https://user-images.githubusercontent.com/36074590/86600263-682c6a00-bfd2-11ea-8630-735595f0619c.png)
![image](https://user-images.githubusercontent.com/36074590/86600352-87c39280-bfd2-11ea-9f66-05412a120036.png)
![image](https://user-images.githubusercontent.com/36074590/86600412-9ad66280-bfd2-11ea-8f97-cc1869a2181b.png)
2. I double-checked the score loss and it surely could be backpropagated to the backbone.

**Comment by nicolas-chaulet** [comment]
Thank you soo much!! No idea what's going on in my end...

**Comment by llijiang** [comment]
I guess the problem may come from the connection parts between backbone features and score net (e.g. the re-voxelization of proposals). Just a guess...

**Comment by nicolas-chaulet** [comment]
Yes looking into that exactly. Because all the rest seems to behave just fine and there must be something really worng to create such a big jump. Thank you so much for looking into it, it's super useful!!

**Comment by nicolas-chaulet** [comment]
Ok, that's fixed, some silly tensor size that was wrong at the end of the scorer getting into the binary cross entropy. I'll let you know if I find anything interesting on my side!

**Comment by llijiang** [comment]
Glad to hear that!

**Comment by PatrickFeng** [comment]
@llijiang Would you like to share all the train/val curves?
